helpers/python

```diff
@@ -3,12 +3,20 @@
 import pkgutil
 import sys
 
-# walk_packages() is much slower than iter_modules(), use it only when
-# completing something with a dot in it.
+
+def iter_submodules(pkgname):
+    __import__(pkgname)
```

Comment:
Some concerns here.

First, the docs of `__import__` say
> Direct use of `__import__()` is also discouraged in favor of `importlib.import_module()`.
...and in pydoc
> ... meant for use by the Python interpreter and not for general use ...

Second, I'm not that happy with importing in the first place, due to possible side effects in doing so, also possible security issues arising from that.

WDYT?

Reply:
> First, the docs of `__import__` say
> 
> > Direct use of `__import__()` is also discouraged in favor of `importlib.import_module()`.

I didn't know what versions of Python `bash-completion` is supposed to support. I decided to stick to the lowest version able to run the current helper. For CPython it seems to be 2.5 and 3.0 (not tested, I only inspected source code of `pkgutil`: [v2.5a2][src-pkgutil-py2], [v3.0a1][src-pkgutil-py3]). On the other hand, `importlib` was introduced in Python 2.7 and 3.1 (docs: [Python 2][doc-importlib-py2], [Python 3][doc-importlib-py3]). So I chose to use `__import__()` instead. Moreover, this is what the implementation of `walk_packages()` uses.

[src-pkgutil-py2]: https://github.com/python/cpython/blob/v2.5a2/Lib/pkgutil.py#L70-L108
[src-pkgutil-py3]: https://github.com/python/cpython/blob/v3.0a1/Lib/pkgutil.py#L71-L150
[doc-importlib-py2]: https://docs.python.org/2/library/importlib.html
[doc-importlib-py3]: https://docs.python.org/3/library/importlib.html

I can switch to using `importlib` if compatibility with old Pythons is not an issue. Just let me know.

> Second, I'm not that happy with importing in the first place, due to possible side effects in doing so, also possible security issues arising from that.

The current script is already doing imports. Quoting [the docs of `walk_packages()`][doc-walk_packages]:

[doc-walk_packages]: https://docs.python.org/3/library/pkgutil.html#pkgutil.walk_packages

> Note that this function must import all packages (not all modules!) on the given path, in order to access the `__path__` attribute to find submodules.

I must admit, though, that my code may in some cases import something that the original version wouldn't. I will modify it to check whether the module I am going to import is a package, in order to match the behavior of the current version. (Although this is non-trivial).

Reply:
Hm.

What i'd ideally like to see is:
1) old python versions not dropped
2) no use of discouraged functions
3) no arbitrary code execution issues out of the box in expected default configs

For 1) and 2) I suppose using importlib if available and __import__ only if not could be the way to go here, unless it gets "too ugly".

For 3) I think it's not actually as bad as I initially thought it'd be, as I don't think we're actually importing anything from the current dir unless the user has explicitly wanted to do that through use of `PYTHONPATH`. So from that POV we're fine with importing stuff in the first place.

Wrt old python versions, I think our CI setup goes down to 2.6.6 on CentOS 6.